### DISCUSSION TITLE
compiling grblHAL RP2040 from source on Linux

### DISCUSSION THREAD

**Discussion by chris2560** [grblHAL/RP2040#79] [Ideas]
I'm struggling to compile grblHAL from source on Linux.

The web builder will not allow me to select the generic 4 axis pinout, which I need because I'm building the circuit board and can't/unwilling to use extra io expanders when there's plenty of pins on the Pico for my simple 4 axis setup.  A PicoCNC board is of no use because I'm controlling a Seig X1, running DM556 stepper drivers with 3NM steppers.  The generic4axis pinout looks good, very similar to the pinout I've been using on a breadboard for the last year, but I'm now adding the 4th axis and need the newer firmware features like hard limits enabled on XYZ but disabled on A.

There is no documentation / I cannot find the documentation for how to compile grblHAL RP2040 from source using cmake and make commands or using Arduino IDE.  I can easily obtain the source code with the git clone command, but cmake wants an SDK AND a tool path chain.  Trying to give it the filepaths to the sdk and tool path chains within my RP2040 module install in Arduino IDE has worked before, but won't work now.  I keep getting __retarget_lock_acquire_recursive errors.

**Comment by terjeio** [comment]
VSCode is the default build environment.

I am using cmake + make for the web builder.
The commands I use are:
```
cd <project folder>
cmake .
make
```
The build process locates the SDK via the `PICO_SDK_PATH` environment variable which has to be set first.
The toolchain should already be on `PATH`?
I installed it with this command:
`apt-get -y install cmake gcc-arm-none-eabi libnewlib-arm-none-eabi `

For the time beeing I am not going to spend time on adding support for the Arduino IDE.

**Reply by chris2560** [reply]
Thanks for the reply, I've tried installing the gcc-arm-none-eabi and libnewlib-arm-none-eabi but they will only install on my older Debian 10 computer (I run Fedora on the newer ones) but the cmake command and I can't then find them.

Seems to have finally managed to compile using the PICO_SDK_FETCH_FROM_GIT=1 environment variable setting on Debian 
I have copied the uf2 to a spare pi pico and seems to connect and run with native usb - (the DRO screen on my interface shows the handwheel can move the A Axis, and programs appear to run fine but not yet wired this spare pico up to the steppers)

However when I tried this command on Fedora, I got:
> arm-none-eabi-gcc: fatal error: cannot read spec file 'nosys.specs': No such file or directory

I'm unsure why it won't work on Fedora when it works on Debian?
(but I'm grateful the option is there to automatically download the SDK and at least works on Debian so I can now get my 4th axis wired up and running)

**Reply by terjeio** [reply]
> I'm unsure why it won't work on Fedora when it works on Debian?

I do not know. And I simply do not have the time to set up, verify and document all possible combinations of development environments.

**Reply by chris2560** [reply]
For some reason my pin definitions are haywire too - the compiler did give warnings of redefinition of pins.  Yes I can understand I'm ending up adding more to your already full schedule by asking for help compiling this myself.

Please can we have the generic_map and generic_4_axis_map provided in the source code as available maps on the web compiler? - (these ones use only pins on the pico board and no extra io expanders)

Thank you for your time in trying to help my mess.

**Reply by terjeio** [reply]
I've now added the generic boards to the Web Builder, not sure if the 3 axis one will compile correctly. Please try and report back.

**Reply by chris2560** [reply]
Thank you,

I have had some chance to do some more testing, on 2 different picos (though not a pico W because I don't own one)

The 3 axis map seems to compile and run fine, and all direction and limit pins are functioning correctly, I haven't yet tried running the machine from it yet, my machine doesn't yet have automatic control of the spindle, the 3 spindle control pins do seem to function correctly when checked with a voltmeter.  I've checked the step, direction and limit pins, and wired my A axis motor into x, y and z and checked they run in forwards and reverse correctly.

The step and limit pins all function correctly on the generic 4 axis map.
But the generic 4 axis map's direction pins seem problematic, on the current firmware version (20231005) and older firmware version from a year ago.  The Y direction controls pin D6 and the z direction outputs to pin D7, the x seems to not have a direction pin and D5 works as the A axis step pin, the source code clearly sets the direction pins as X=D6, Y=D7, Z=D8, A=Z+1=D9.

When the 3 axis pinout's direction pins are set to x=D7, y=D5, z=D6 it doesn't work at all, nor when the 4 axis map's direction pins are set to x=D7, y=D8, z=D9.

Is there some set of rules which define what pins can be used for what and in what order? (similar to the strict rules on GRBL when I was setting up GRBL on an Atmega1284 2 years ago)?

**Reply by terjeio** [reply]
There are strict rules depending on how pins are written to.
E.g. `#define DIRECTION_OUTMODE       GPIO_SHIFT5` uses a shift and writes all direction pins in one go. BTW this should be GPIO_SHIFT6 if I am not mistaken, it would explain the mismatch described above.

Other options for writing to pins or a set of pins are [defined here](https://github.com/grblHAL/RP2040/blob/1c4ec54fc6ec2d4f18a1b02ef65c2122f028659b/driver.h#L61-L100), with variyng degree of freedom to chose pin numbers.

Note that step pins are written by PIO code in the 4 axis map - which requires the pins to be consecutive.

**Reply by chris2560** [reply]
That's solved it.
Changing line 40 of generic_4axis_map.h
from 

> #define DIRECTION_OUTMODE       GPIO_SHIFT5

to
 
> #define DIRECTION_OUTMODE       GPIO_SHIFT6

and once compiled, the step, direction and limit pins all function correctly.

I'd seen the shifts defined in drive.h but not understood what they do, but basically the 

> #define X/Y/Z_DIRECTION_PIN      6/7/8

lines have no effect?

Please can line 40 of generic_4axis_map.h be changed on the repo accordingly?

Many thanks

**Reply by terjeio** [reply]
The pin definitions are only for reference and for generationg the $pins command output when a shift is used.
The repo and the Web Builder has now been updated.

**Reply by chris2560** [reply]
Many thanks,
The web builder now compiles the generic 4th axis map really well - (The version I'd compiled which used native USB seemed to weirdly disconnect from serial every time I started all 3 XYZ axis moving at once or maybe when I spun the spindle up to 5000rpm, but this was fixed with re-compiling for UART and using my FTDI UART adaptor - I suspect it may have been my system overloading some IO pins or a dodgy wire disconnecting from the laptop when the workbench started vibrating as the spindle spun up)

**Comment by chris2560** [comment]
Should my 4th axis home just like the XYZ axis?
I've got an infrared beam break for a limit switch and 2 drive transistors, connecting via 50ohms and 1microFarad capacitance to the IO pin, just like the other axis.
When the A axis motor and limit switch is plugged into the Z axis pins, it homes just fine, however when plugged into the A axis pins and $HA is called, it simply zero's the machine coordinate A immediately and doesn't seem to move the A axis either direction as it does.  

> [MSG:] 

is printed to the console when $HA is sent.

I can post the parameters here if necessary.

**Reply by terjeio** [reply]
Have you enabled A-axis homing? $44-$47 are the settings for the different passes, and one of the passes should have the A-Axis bit set. E.g. $47=8 for homing A separately.

**Reply by chris2560** [reply]
You've solved my trouble yet again, it now tries to home the A axis when I send $Ha.  (just a shame my infrared limit switch is not working reliably now,  but that's my problem, not any problem with the software)
Many Thanks
